### PR TITLE
macOS Preferences: require restart for new skin or new overview type

### PR DESCRIPTION
**Pull request by ronso0**
supposed to work around [Skin switch crash on OSX](https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/Skin.20switch.20crash.20on.20OSX)

for the [Linux crash](https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/skin.20switch.20crash.20on.20Linux) #3144 is necessary, too, because on Linux Deere would also crash on shutdown and drop users' Deere configuration.
Didn't test if #3192 has any effect on that.

"reload skin" is left in Developer menu.

### CODE REVIEW

**Comment by Be-ing** [comment]
This annoying limitation should only be imposed where it is necessary. IIUC that would only be macOS. I don't think we have a way to predict which macOS machines will crash, so I'd be okay with requiring a restart for all of them, but requiring a restart for Windows and Linux would be annoying an unnecessary.

**Comment by ronso0** [comment]
so `#ifndef __APPLE__` for the automatic skin reload? Is that the correct way?

**Comment by Be-ing** [comment]
yes

**Comment by ronso0** [comment]
Ready for review.

File: src/preferences/dialog/dlgprefwaveform.cpp [hunk 1]
```diff
@@ -227,6 +236,12 @@ void DlgPrefWaveform::slotResetToDefaults() {
     playMarkerPositionSlider->setValue(50);
 }
 
+void DlgPrefWaveform::notifyRebootNecessary() {
+    // make the fact that you have to restart mixxx more obvious
+    QMessageBox::information(
+            this, tr("Information"), tr("Mixxx must be restarted to load the new overview type."));
```
Comment (Be-ing) [inline src/preferences/dialog/dlgprefwaveform.cpp:242]: This is used both for changing the skin and overview types.

Comment (ronso0) [reply src/preferences/dialog/dlgprefwaveform.cpp:242]: ?
the skin and locale are set in src/preferences/dialog/dlgprefinterface.cpp

Comment (Be-ing) [reply src/preferences/dialog/dlgprefwaveform.cpp:242]: Read the text here. It only mentions the overview type.

Comment (ronso0) [reply src/preferences/dialog/dlgprefwaveform.cpp:242]: yes, read the file name please :)
we're in src/preferences/dialog/dlgpref**waveform**.cpp and here the only option that requires a skin reload is the overview.

Comment (ronso0) [reply src/preferences/dialog/dlgprefwaveform.cpp:242]: src/preferences/dialog/dlgpref**interface**.cpp has different messages

Comment (Be-ing) [reply src/preferences/dialog/dlgprefwaveform.cpp:242]: Okay. Sorry for the noise.

File: CHANGELOG.md [hunk 2]
```diff
@@ -33,6 +33,7 @@
 * Add laptop battery widget to skins [#2283](https://github.com/mixxxdj/mixxx/pull/2283) [#2277](https://github.com/mixxxdj/mixxx/pull/2277) [#2250](https://github.com/mixxxdj/mixxx/pull/2250) [#2228](https://github.com/mixxxdj/mixxx/pull/2228) [#2221](https://github.com/mixxxdj/mixxx/pull/2221) [#2163](https://github.com/mixxxdj/mixxx/pull/2163) [#2160](https://github.com/mixxxdj/mixxx/pull/2160) [#2147](https://github.com/mixxxdj/mixxx/pull/2147) [#2281](https://github.com/mixxxdj/mixxx/pull/2281) [#2319](https://github.com/mixxxdj/mixxx/pull/2319) [#2287](https://github.com/mixxxdj/mixxx/pull/2287)
 * Show when passthrough mode is active on overview waveforms [#2575](https://github.com/mixxxdj/mixxx/pull/2575) [#2616](https://github.com/mixxxdj/mixxx/pull/2616)
 * Changed format of currently playing track in window title from "artist, title" to "artist - title" [#2807](https://github.com/mixxxdj/mixxx/pull/2807)
+* Workaround macOS skin change crash by requiring Mixxx restart [#3252](https://github.com/mixxxdj/mixxx/pull/3252) [lp:1877487](https://bugs.launchpad.net/mixxx/+bug/1877487)
```
Comment (Be-ing) [inline CHANGELOG.md:36]: AFAIK the crash is new in 2.3 so I don't think it should be mentioned in the changelog.

File: src/preferences/dialog/dlgprefinterface.cpp [hunk 3]
```diff
@@ -403,15 +409,19 @@ void DlgPrefInterface::slotApply() {
...
 
     if (locale != m_localeOnUpdate) {
-        notifyRebootNecessary();
+        notifyLocaleRebootNecessary();
         // hack to prevent showing the notification when pressing "Okay" after "Apply"
         m_localeOnUpdate = locale;
     }
 
     if (m_bRebootMixxxView) {
+#ifdef __APPLE__
```
Comment (Be-ing) [inline src/preferences/dialog/dlgprefinterface.cpp:418]: add link to https://bugs.launchpad.net/mixxx/+bug/1877487 in a comment

File: CHANGELOG.md (hunk 2 above)
Comment (ronso0) [reply CHANGELOG.md]: alright.

**Comment by ronso0** [comment]
the only thing is that there are now two notifications in Interface: one for a changed locale, one for the changed skin.
Considering users change the locale only once this is bearable.

**Comment by Be-ing** [comment]
The Travis build was cancelled?

**Comment by ronso0** [comment]
@Be-ing Conflicts are resolved, can you give this a try on your mac?

**Comment by Be-ing** [comment]
I can test this but I never encountered the skin switch crash, so I can't confirm if this successfully works around it.

**Comment by ronso0** [comment]
that'll be sufficient, I just want to be sure the skin switching works as expected.
after merge mac users can test with the new builds.

**Comment by ronso0** [comment]
I already tested with @steve that the current skins don't crash on shutdown.

**Comment by Be-ing** [comment]
I confirm this works as intended. Whether it actually works around the crash I still do not know because I never reproduced the crash. I am reluctant to bother all macOS users with this annoyance without confirmation that it actually works around the crash. After we merge #3305, you can merge 2.3 to this branch and affected macOS users can easily test the artifact from GitHub Actions.

**Comment by ronso0** [comment]
Sure. I'd love to apply that workaround only to macOS and affected Qt versions if we can track those down.

**Comment by Holzhaus** [comment]
Could we check if the crash is not fixed by the new CI/buildenv? Maybe it was some obscure compiler bug.

**Comment by Be-ing** [comment]
Or it might be fixed with the new build environment.

**Comment by Be-ing** [comment]
Workaround confirmed by @hile.

**Pull request merged**